src/hotspot/share/utilities/debug.hpp

```diff
@@ -46,6 +46,98 @@ bool handle_assert_poison_fault(const void* ucVoid, const void* faulting_address
+// those operations are not normally permitted, with the state checked by an
+// assert.  We want the debugging commands to bypass those checks.
+class DebuggingContext {
+  static int _enabled;          // Nesting counter.
+
+public:
+  DebuggingContext();
+  ~DebuggingContext();
+  // Asserts and other code use this to determine whether to bypass checks
+  // that would otherwise lead to program termination.
+  static bool is_enabled() { return _enabled > 0; }
+};
+
+// VMASSERT_CHECK_PASSED(P) provides the mechanism by which DebuggingContext
+// disables asserts.  It returns true if P is true or DebuggingContext is
+// enabled.  Assertion failure is reported if it returns false, terminating
+// the program.
+//
+// The DebuggingContext check being enabled isn't placed inside the report
+// function, as that would prevent the report function from being noreturn.
+// The report function should be noreturn so there isn't a control path to the
+// assertion's continuation that has P being false.  Otherwise, the compiler
+// might logically split the continuation to include that path explicitly,
+// possibly leading to discovering (and warning about) invalid code.  For
+// example, if P := x != nullptr, and the continuation contains a dereference
+// of x, the compiler might warn because there is a control path (!P -> report
+// -> continuation) where that dereference is known to be invalid.  (Of
+// course, if execution actually took that path things would go wrong, but
+// that's the risk the DebuggingContext mechanism takes.)
+//
+// Similarly, the check for enabled DebuggingContext shouldn't follow P.
+// Having this macro expand to `P || DebuggingContext::is_enabled()` has the
+// same problem of a control path through !P to the assertion's continuation.
+//
+// But it can't be just `DebuggingContext::is_enabled() || P` either.  That
+// prevents the compiler from inferring based on P that it is true in the
+// continuation.  But it also prevents the use of assertions in constexpr
+// contexts, since that expression is not constexpr.
+//
+// We could accomodate constexpr usage with std::is_constant_evaluated() (from
+// C++20). Unfortunately, we don't currently support C++20.  However, most
+// supported compilers have implemented it, and that implementation uses a
+// compiler intrinsic that we can use directly without otherwise using C++20.
+//
+// Note that if we could use std::is_constant_evaluated() then we could just
+// use this definition for DebuggingContext::is_enabled:
+//   static constexpr bool is_enabled() {
+//     return !std::is_constant_evaluated() && _enabled;
+//   }
+// The idea being that we are definitely not executing for debugging if doing
+// constant evaluation in the compiler. We don't do something like that now,
+// because we need a fallback when we don't have any mechanism for detecting
+// constant evaluation.
+#if defined(TARGET_COMPILER_gcc) || defined(TARGET_COMPILER_xlc)
```

Comment:
All this seems like it should go in COMPILER_HEADER(globalDefinitions.hpp) but since globalDefinitions.hpp includes debug.hpp, so you can't do this.  Can we file an RFE to clean this up (if possible)?

Reply:
I considered that, and had two reasons for rejecting it.

The include ordering problem was one. There are uses of assert and the like in
globalDefinitions. Some of those might someday be moved elsewhere, but for now
the order is a problem.

The other is that this is really not "general purpose", but rather very
specifically tailored to the use here.  And it can't be general purpose,
because there isn't a general fallback for testing for constexpr evaluation.

I could have added `HAS_IS_CONSTANT_EVALUATED` and ::is_constant_evaluated(),
with all uses of the latter requiring protection by the former. But I couldn't
think of any places where I would want to use that. Using it in, for example,
count_trailing_zeros to allow it to be (sometimes) constexpr would make the
question of whether it's constexpr be platform-dependent, which means it can't
be used in shared required constexpr contexts. And it's even worse since the
platform dependency is also currently compiler version dependent. I think that
path leads nowhere good.

However, I've filed this cleanup RFE:
https://bugs.openjdk.org/browse/JDK-8303797